pkg/handlers/handler_base.go

```diff
@@ -0,0 +1,171 @@
+package handlers
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/facebookgo/clock"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+
+	"github.com/cmsgov/easi-app/pkg/appcontext"
+	"github.com/cmsgov/easi-app/pkg/apperrors"
+)
+
+// NewHandlerBase is a constructor for HandlerBase
+func NewHandlerBase(logger *zap.Logger) HandlerBase {
+	return HandlerBase{
+		logger: logger,
+		clock:  clock.New(),
+	}
+}
+
+// HandlerBase is for shared handler utilities
+type HandlerBase struct {
+	logger *zap.Logger
+	clock  clock.Clock
+}
+
+type errorItem struct {
+	Field   string `json:"field"`
+	Message string `json:"message"`
+}
+
+// errorResponse contains the structure of error for a http response
+type errorResponse struct {
+	Errors  []errorItem `json:"errors"`
+	Code    int         `json:"code"`
+	Message string      `json:"message"`
+	TraceID uuid.UUID   `json:"traceID"`
+}
+
+func newErrorResponse(code int, message string, traceID uuid.UUID) errorResponse {
+	return errorResponse{
+		Errors:  []errorItem{},
+		Code:    code,
+		Message: message,
+		TraceID: traceID,
+	}
+}
+
+func (r *errorResponse) withMap(errMap map[string]string) {
+	for k, v := range errMap {
+		r.Errors = append(r.Errors, errorItem{
+			Field:   k,
+			Message: v,
+		})
+	}
+}
+
+// WriteErrorResponse writes a response for a given application error
+func (b HandlerBase) WriteErrorResponse(ctx context.Context, w http.ResponseWriter, appErr error) {
+	logger, ok := appcontext.Logger(ctx)
+	if !ok {
+		logger = b.logger
+	}
+
+	traceID, ok := appcontext.Trace(ctx)
+	if !ok {
+		traceID = uuid.New()
+		logger.With(zap.String("traceID", traceID.String()))
+	}
+
+	// get code and response
+	var code int
+	var response errorResponse
+	switch appErr := appErr.(type) {
+	case *apperrors.UnauthorizedError:
+		// 4XX errors are not logged as errors, but are for client
+		logger.Info("Returning unauthorized response from handler", zap.Error(appErr))
+		code = http.StatusUnauthorized
+		response = newErrorResponse(
+			http.StatusUnauthorized,
+			"Unauthorized",
+			traceID,
+		)
+	case *apperrors.QueryError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Something went wrong",
+			traceID,
+		)
+	case *apperrors.NotificationError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
```

Comment:
I think this should be a 502 Bad Gateway error.

Reply:
Bad gateway is for proxies and caches I believe. I tried to think of something better for this one, but not sure what an auxiliary action should return. I even though about just returning a 200, because the resource was processed just fine.